File: frontend/packages/api-client/src/models/spread.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
import { Page } from "./pagination";

export type SpreadType = "system" | "custom";

export interface SpreadPosition {
  index: number;
  label: string;
  x: number;
  y: number;
  rotation: number;
  scale: number;
}

export interface Spread {
  id: string;
  name: string;
  description: string | null;
  num_cards: number;
  positions: SpreadPosition[];
  prompts: string[];
  allow_reversed: boolean;
  user_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface AdminSpread extends Spread {
  owner_username: string | null;
}

export interface CreateSpreadPayload {
  name: string;
  description?: string | null;
  positions: SpreadPosition[];
  prompts?: string[];
  allow_reversed?: boolean;
}

export interface UpdateSpreadPayload {
  name?: string;
  description?: string | null;
  positions?: SpreadPosition[];
  prompts?: string[];
  allow_reversed?: boolean;
}

export type PaginatedSpreads = Page<AdminSpread>;

File: backend/app/schemas/spread.py
import enum
import uuid
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, field_validator

Prompt = Annotated[str, Field(min_length=1, max_length=200)]


class SpreadType(enum.StrEnum):
    SYSTEM = "system"
    CUSTOM = "custom"


class SpreadPosition(BaseModel):
    index: int = Field(ge=0, le=8)
    label: str = Field(min_length=1, max_length=50)


def _check_unique_indices(positions: list[SpreadPosition] | None) -> list[SpreadPosition] | None:
    if positions is None:
        return positions
    indices = [p.index for p in positions]
    if len(indices) != len(set(indices)):
        raise ValueError("Position indices must be unique")
    return positions


class SpreadCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    positions: list[SpreadPosition] = Field(min_length=1, max_length=9)
    prompts: list[Prompt] = Field(default_factory=list, max_length=10)
    allow_reversed: bool = True

    @field_validator("positions")
    @classmethod
    def unique_indices(cls, positions: list[SpreadPosition]) -> list[SpreadPosition]:
        return _check_unique_indices(positions)


class SpreadUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    positions: list[SpreadPosition] | None = Field(default=None, min_length=1, max_length=9)
    prompts: list[Prompt] | None = Field(default=None, max_length=10)
    allow_reversed: bool | None = None

    @field_validator("positions")
    @classmethod
    def unique_indices(cls, positions: list[SpreadPosition] | None) -> list[SpreadPosition] | None:
        return _check_unique_indices(positions)


class SpreadRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    description: str | None
    num_cards: int
    positions: list[SpreadPosition]
    prompts: list[str]
    allow_reversed: bool
    user_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime


class AdminSpreadRead(SpreadRead):
    owner_username: str | None = None
